harvester/DBInterface.py

```diff
@@ -343,198 +413,150 @@ def write_record(self, record, repo_id, metadata_prefix, domain_metadata):
 
 			if "subject" in record:
-				try:
-					cur.execute(self._prep("DELETE from subjects where record_id = ?"), (record["record_id"],))
-				except:
-					pass
 				if not isinstance(record["subject"], list):
 					record["subject"] = [record["subject"]]
+				existing_subject_ids = self.get_multiple_record_ids("records_x_subjects", "subject_id", "record_id", record["record_id"])
 				for subject in record["subject"]:
-					if (subject == "form_descriptors"):
-						continue
-					try:
-						if subject is not None and len(subject) > 0:
-							cur.execute(self._prep("INSERT INTO subjects (record_id, subject) VALUES (?,?)"),
-										(record["record_id"], subject))
-					except self.dblayer.IntegrityError as e:
-						self.logger.error("Record insertion problem: %s" %e)
-						pass
+					subject_id = self.get_single_record_id("subjects", subject)
+					if subject_id is None:
+						subject_id = self.insert_related_record("subjects", subject)
+					if subject_id is not None:
+						if subject_id not in existing_subject_ids:
+							self.insert_cross_record("records_x_subjects", "subjects", subject_id, record["record_id"])
 
 			if "publisher" in record:
-				try:
-					cur.execute(self._prep("DELETE from publishers where record_id = ?"), (record["record_id"],))
-				except:
-					pass
 				if not isinstance(record["publisher"], list):
 					record["publisher"] = [record["publisher"]]
+				existing_publisher_ids = self.get_multiple_record_ids("records_x_publishers", "publisher_id", "record_id", record["record_id"])
 				for publisher in record["publisher"]:
-					try:
-						if publisher is not None and len(publisher) > 0:
-							cur.execute(self._prep("INSERT INTO publishers (record_id, publisher) VALUES (?,?)"),
-										(record["record_id"], publisher))
-					except self.dblayer.IntegrityError as e:
-						self.logger.error("Record insertion problem: %s" %e)
-						pass
+					publisher_id = self.get_single_record_id("publishers", publisher)
+					if publisher_id is None:
+						publisher_id = self.insert_related_record("publishers", publisher)
+					if publisher_id is not None:
+						if publisher_id not in existing_publisher_ids:
+							self.insert_cross_record("records_x_publishers", "publishers", publisher_id, record["record_id"])
 
 			if "rights" in record:
-				try:
-					cur.execute(self._prep("DELETE from rights where record_id = ?"), (record["record_id"],))
-				except:
-					pass
 				if not isinstance(record["rights"], list):
 					record["rights"] = [record["rights"]]
+				existing_rights_ids = self.get_multiple_record_ids("records_x_rights", "rights_id", "record_id", record["record_id"])
 				for rights in record["rights"]:
-					try:
-						cur.execute(self._prep("INSERT INTO rights (record_id, rights) VALUES (?,?)"),
-									(record["record_id"], rights))
-					except self.dblayer.IntegrityError:
-						pass
+					sha1 = hashlib.sha1()
```

Comment:
Wouldn't mind seeing this commented inline as to why it needs a hash for uniqueness unlike the others -- I get it but it still stands out as different